compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1951,7 +1933,7 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
       val t2 = distributeAnd(tp2, tp1)
       if (t2.exists) t2
       else if (isErased) erasedGlb(tp1, tp2, isJava = false)
-      else liftIfHK(tp1, tp2, op, original)
+      else liftIfHK(tp1, tp2, op, original, _ | _)
```

Comment:
I don't get why we're doing something different with the variance depending on whether we have an AndType or an OrType, the variance position is the same. If I intersect `[+X] => Cov[X]` and `[-X] => Contra[X]` I get an invariant type parameter structurally: `[X] => Cov[X] & Contra[X]`, but it looks like this code would make the type parameter bivariant instead.

Reply:
It's because conceptually variances are attached to abstract types, not the lambdas. Say you have two abstract members
```
trait A { def F[-X] }
trait B { def F[+X] }
object O extends A, B { ... }
```
Here, the only legal implementation of `F` in `O` is indeed bivariant.

I have added a comment to the code.